wallet/walletrpc.c

```diff
@@ -369,13 +369,22 @@ static void json_listfunds(struct command *cmd, const char *buffer UNUSED,
 	struct peer *p;
 	struct utxo **utxos =
 	    wallet_get_utxos(cmd, cmd->ld->wallet, output_state_available);
+	char* out;
+	struct ripemd160 h160;
+	struct pubkey funding_pubkey;
 	json_object_start(response, NULL);
 	json_array_start(response, "outputs");
 	for (size_t i = 0; i < tal_count(utxos); i++) {
 		json_object_start(response, NULL);
 		json_add_txid(response, "txid", &utxos[i]->txid);
 		json_add_num(response, "output", utxos[i]->outnum);
 		json_add_u64(response, "value", utxos[i]->amount);
+		bip32_pubkey(cmd->ld->wallet->bip32_base, &funding_pubkey,
+			     utxos[i]->keyindex);
+		pubkey_to_hash160(&funding_pubkey, &h160);
+		out = p2sh_to_base58(cmd,
```

Comment:
Are all outputs P2SH-P2WPKH? I think some outputs are P2WPKH now?

Reply:
Oh sorry, https://github.com/lightningnetwork/lightning-rfc/blob/master/03-transactions.md#funding-transaction-output  said that the funding output is P2WSH. Do you mean this one?

Reply:
This code is inside the `outputs` array loop describing onchain addresses, not the funding transaction of `channels`.  My understanding is that we recently added the ability to receive onchain via P2WPKH addresses.  Also I think change addresses are P2WPKH.

Reply:
See `enum wallet_output_type` in `wallet/wallet.h`.  Various outputs have different types that change the effective onchain address they will have.

Reply:
@ZmnSCPxj if so, then maybe it is worth to add such `type` to `utxo` struct?

Reply:
Maybe `is_p2sh` and `close_info` are enough to derive information on exact type of address? Will need to trace what `is_p2sh` means at least (does it differentiate between P2SH-P2WPKH vs P2WPKH?)

Reply:
Investigating `wallet_stmt2output` in `wallet/wallet.c`, yes, `is_p2sh` is enough to differentiate P2SH-P2WPKH from the rest.  Probably `close_info` is for outputs that are not yet claimable (they need onchaind to claim it I think...?), so suppressing output of address field may be OK for outputs with non-NULL `close_info`, but I have not investigated deeply.